jdaviz/core/astrowidgets_api.py

```diff
@@ -0,0 +1,523 @@
+import sys
+
+import numpy as np
+from astropy import units as u
+from astropy.coordinates import SkyCoord
+from astropy.utils.introspection import minversion
+from astropy.wcs import NoConvergence
+from echo import delay_callback
+from glue.config import colormaps
+from glue.core import Data
+
+from jdaviz.configs.imviz.helper import data_has_valid_wcs, get_top_layer_index
+from jdaviz.core.events import SnackbarMessage
+
+__all__ = ['AstrowidgetsImageViewerMixin']
+
+ASTROPY_LT_4_3 = not minversion('astropy', '4.3')
+
+
+class AstrowidgetsImageViewerMixin:
+    """This class implements ``astrowidgets`` API for Jdaviz *image* viewer.
+    This does not provide a fully functional viewer, but rather
+    should be used as mixin into an existing viewer subclass.
+
+    A child class that uses this must run :meth:`init_astrowidgets_api`
+    within its ``__init__``.
+
+    .. note:: Do not use this for spectral or cube viewers.
+
+    """
+    RESERVED_MARKER_SET_NAMES = ['all']
+
+    # __init__ not called, so use this to setup.
+    def init_astrowidgets_api(self):
+        """This method must be called in child class ``__init__``."""
+        # Markers
+        self._marktags = set()
+        self._default_mark_tag_name = 'default-marker-name'
+        # marker shape not settable: https://github.com/glue-viz/glue/issues/2202
+        self.marker = {'color': 'red', 'alpha': 1.0, 'markersize': 5}
+
+    def save(self, filename):
+        """Save out the current image view to given PNG filename."""
+        if not filename.lower().endswith('.png'):
+            filename = filename + '.png'
+        self.figure.save_png(filename=filename)

```

Comment:
codecov says this isn't covered by the tests.  Is that intentional? I guess it's in the notebook... but maybe it can be added now if it's easy? (if not easy, not critical!)

Reply:
Funny you asked. I actually wrote a test but I cannot run it due to bqplot/bqplot#1393 .

https://github.com/spacetelescope/jdaviz/blob/5d8a896e1249f643e2db27340c1597882acb9fb2/jdaviz/configs/imviz/tests/test_astrowidgets_api.py#L13-L15